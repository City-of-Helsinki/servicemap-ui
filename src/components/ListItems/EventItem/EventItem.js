/* eslint-disable camelcase */
import React from 'react';
import PropTypes from 'prop-types';
import { intlShape } from 'react-intl';
import { Event } from '@material-ui/icons';
import ResultItem from '../ResultItem';

const formatEventDate = (event, intl) => {
  const timeString = intl.formatMessage({ id: 'general.time.short' });
  const start = new Date(event.start_time);
  const end = new Date(event.end_time);
  const startDate = intl.formatDate(start, { month: 'numeric', day: 'numeric' });
  const endDate = intl.formatDate(end, { month: 'numeric', day: 'numeric' });
  const startTime = `${timeString} ${intl.formatTime(start)}`;
  const startWeekDay = intl.formatDate(start, { weekday: 'short' });

  if (startDate !== endDate) {
    return `${startDate}â€“${endDate}`;
  }
  let dateString = `${startWeekDay} ${startDate} ${startTime}`;
  // Check how many days to start of event. Returns number of days or "today"/"tomorrow"
  const daysTo = intl.formatRelative(start, { units: 'day' });
  // Check if string has numbers. If false, means that value is either "today" or "tomorrow"
  if (!/\d/.test(daysTo)) {
    // Instead of date, display "today" or "tomorrow"
    dateString = `${daysTo} ${startTime}`;
  }
  return dateString;
};

const EventItem = ({
  changeSelectedEvent,
  event,
  getLocaleText,
  intl,
  navigator,
}) => {
  const dateString = formatEventDate(event, intl);
  return (
    <ResultItem
      key={event.id}
      icon={<Event />}
      title={getLocaleText(event.name)}
      subtitle={dateString}
      divider
      onClick={(e) => {
        e.preventDefault();
        if (navigator) {
          changeSelectedEvent(event);
          navigator.push('event', event.id);
        }
      }}
    />
  );
};

export default EventItem;

// Typechecking
EventItem.propTypes = {
  changeSelectedEvent: PropTypes.func.isRequired,
  event: PropTypes.objectOf(PropTypes.any).isRequired,
  getLocaleText: PropTypes.func.isRequired,
  intl: intlShape.isRequired,
  navigator: PropTypes.objectOf(PropTypes.any),
};

EventItem.defaultProps = {
  navigator: null,
};
