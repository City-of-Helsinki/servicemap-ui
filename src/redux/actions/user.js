import LocalStorageUtility from '../../utils/localStorage';
import fetchAddress from '../../views/MapView/utils/fetchAddress';

export const setLocale = locale => ({
  type: 'SET_LOCALE',
  locale,
});

export const setInitialLoad = () => ({
  type: 'SET_INITIAL_LOAD',
});

export const setPage = page => ({
  type: 'SET_CURRENT_PAGE',
  page,
});

export const setUserPosition = position => ({
  type: 'SET_POSITION',
  position,
});

const setTheme = theme => ({
  type: 'SET_THEME',
  theme,
});


export const changeLocaleAction = locale => async (dispatch) => {
  dispatch(setLocale(locale));
};

export const actionSetInitialLoad = () => async (dispatch) => {
  dispatch(setInitialLoad());
};

export const setCurrentPage = page => async (dispatch) => {
  dispatch(setPage(page));
};

export const changeTheme = theme => async (dispatch) => {
  LocalStorageUtility.saveItem('theme', theme);
  dispatch(setTheme(theme));
};

export const findUserLocation = () => async (dispatch) => {
  const success = (position) => {
    if (position.coords.accuracy < 10000) {
      fetchAddress({ lat: position.coords.latitude, lng: position.coords.longitude })
        .then((data) => {
          dispatch(setUserPosition({
            coordinates: position.coords,
            allowed: true,
            addressData: data,
          }));
        });
    } else {
      console.warn(`Position accuracy: ${position.coords.accuracy}. Max accuracy: 10000`);
      dispatch(setUserPosition({ coordinates: null, allowed: true, addressData: null }));
    }
  };

  const error = (err) => {
    console.warn(`GeoLocation error:(${err.code}): ${err.message}`);
    if (err.code === 1) {
      dispatch(setUserPosition({ coordinates: null, allowed: false, addressData: null }));
    } else {
      dispatch(setUserPosition({ coordinates: null, allowed: true, addressData: null }));
    }
  };

  navigator.geolocation.getCurrentPosition(success, error);
};


export default { setLocale, changeLocaleAction };
