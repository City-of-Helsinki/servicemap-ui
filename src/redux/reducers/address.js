const initialState = {
  units: [],
  addressCoordinates: null,
  addressData: null,
  adminDistricts: [],
  toRender: 'adminDistricts',
};

export default (state = initialState, action) => {
  switch (action.type) {
    case 'SET_ADDRESS_DATA':
      return {
        ...state,
        addressData: action.data,
      };
    case 'SET_ADDRESS_LOCATION':
      return {
        ...state,
        addressCoordinates: action.location.addressCoordinates,
      };
    case 'SET_ADDRESS_UNITS':
      return {
        ...state,
        units: action.units,
      };
    case 'SET_ADMINISTRATIVE_DISTRICTS':
      return {
        ...state,
        adminDistricts: action.data,
      };
    case 'SET_TO_RENDER':
      return {
        ...state,
        toRender: action.data,
      };
    default:
      return state;
  }
};
