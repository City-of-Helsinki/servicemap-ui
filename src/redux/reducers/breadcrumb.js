const initialState = [];

const breadcrumbReducer = (state = initialState, action) => {
  switch (action.type) {
    case 'PUSH_ENTRY':
      return [...state, action.entry];
    case 'POP_ENTRY': {
      const newEntries = Array.from(state);
      newEntries.pop();
      return newEntries;
    }
    case 'REPLACE_ENTRY': {
      const newEntries = Array.from(state);
      newEntries.pop();
      return [...newEntries, action.entry];
    }
    default:
      return state;
  }
};

export default breadcrumbReducer;
