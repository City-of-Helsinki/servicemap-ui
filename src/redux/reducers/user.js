
const initialState = {
  initialLoad: false,
  locale: 'fi',
  page: 'home',
  theme: 'default',
  position: {
    coordinates: null,
    allowed: false,
  },
};

export default (state = initialState, action) => {
  switch (action.type) {
    case 'SET_LOCALE':
      return {
        ...state,
        locale: action.locale,
      };
    case 'SET_INITIAL_LOAD':
      return {
        ...state,
        initialLoad: true,
      };
    case 'SET_CURRENT_PAGE':
      return {
        ...state,
        page: action.page,
      };
    case 'SET_POSITION':
      return {
        ...state,
        position: action.position,
      };
    case 'SET_THEME':
      return {
        ...state,
        theme: action.theme,
      };
    default:
      return state;
  }
};
