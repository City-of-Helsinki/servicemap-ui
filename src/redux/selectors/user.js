
export const getTheme = state => state.user.theme;
export const getPage = state => state.user.page;
export const getLocale = state => state.user.locale;
