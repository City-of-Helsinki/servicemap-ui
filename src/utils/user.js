import { useSelector } from 'react-redux'
 
export const useUserTheme = () => useSelector(state => state.user.theme);
export const useUserLocale = () => useSelector(state => state.user.locale);
