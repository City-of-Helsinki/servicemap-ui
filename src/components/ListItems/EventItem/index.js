
import { connect } from 'react-redux';
import { injectIntl } from 'react-intl';
import EventItem from './EventItem';
import { getLocaleString } from '../../../redux/selectors/locale';
import { changeSelectedEvent } from '../../../redux/actions/event';

// Listen to redux state
const mapStateToProps = (state) => {
  const getLocaleText = textObject => getLocaleString(state, textObject);
  const {
    navigator,
  } = state;
  return {
    getLocaleText,
    navigator,
  };
};

export default injectIntl(connect(
  mapStateToProps,
  { changeSelectedEvent },
)(EventItem));
