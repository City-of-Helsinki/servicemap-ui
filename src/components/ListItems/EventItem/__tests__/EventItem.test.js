import React from 'react';
import { initialState } from '../../../../redux/reducers/user';
import EventItem from '../index';
import { getRenderWithProviders } from '../../../../../jestUtils';

const mockData = {
  "name": {
      "fi": "Tapahtuma kerho",
      "sv": "Tapahtuma kerho sv",
      "en": "Tapahtuma kerho en"
  },
  "location": {
    "id": 111,
    "name": {
      "fi": "Tapahtuma kerhon paikka",
      "sv": "Tapahtuma kerhon paikka sv",
      "en": "Tapahtuma kerhon paikka en"
    },
  },
  "start_time": "2022-08-31T18:00:00Z",
  "end_time": "2022-08-31T18:00:00Z",
  "id": 813,
};

// Generic required props for SimpleListItem
const mockProps = {
  event: mockData,
};




const renderWithProviders = getRenderWithProviders({
  user: initialState,
  settings: {},
  service: {
    current: null,
  },
});

describe('<EventItem />', () => {

  it('does render accessibility attributes correctly', () => {
    const { container } = renderWithProviders(<EventItem {...mockProps} />);
    const items = container.querySelectorAll('li');
    const firstItem = items[0];
    const firstItemResultTitle = firstItem.querySelectorAll('p')[0];
    const dividerItem = items[1];

    // List item's image should be aria-hidden
    expect(firstItem.querySelector('svg').getAttribute('aria-hidden')).toBeTruthy();
    expect(firstItem.getAttribute('role')).toEqual('link');
    expect(firstItem.getAttribute('tabIndex')).toEqual('0');
    expect(firstItemResultTitle.getAttribute('aria-hidden')).toBeFalsy();
    expect(dividerItem.getAttribute('aria-hidden')).toBeTruthy();
  });
});
