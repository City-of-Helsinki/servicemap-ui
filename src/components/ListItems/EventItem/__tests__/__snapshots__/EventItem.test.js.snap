// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EventItem /> should work 1`] = `
<div>
  <li
    class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-button ResultItem-listItem-2 css-16e6fe9-MuiButtonBase-root-MuiListItem-root"
    role="link"
    tabindex="0"
  >
    <div
      class="MuiListItemIcon-root ResultItem-listItemIcon-3 css-cveggr-MuiListItemIcon-root"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-6yqi1c-MuiSvgIcon-root"
        data-testid="EventIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
        />
      </svg>
    </div>
    <div
      class="ResultItem-itemTextContainer-9  "
    >
      <div
        class="ResultItem-topRow-11"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 ResultItem-title-7 ResultItem-srOnly css-1ak20dk-MuiTypography-root"
          style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
        >
          
    Tapahtuma kerho 
    Tapahtuma kerhon paikka 
     
    Wed 8/31 general.time.short 9:00 PM 
    
  
        </p>
        <p
          aria-hidden="true"
          class="MuiTypography-root MuiTypography-body2 ResultItem-title-7    ResultItem-title css-1ak20dk-MuiTypography-root"
          role="textbox"
        >
          Tapahtuma kerho
        </p>
      </div>
      <div>
        <div>
          <p
            aria-hidden="true"
            class="MuiTypography-root MuiTypography-caption ResultItem-noMargin-8 ResultItem-text-16  ResultItem-subtitle css-r3hlsg-MuiTypography-root"
          >
            Tapahtuma kerhon paikka
          </p>
        </div>
        <div
          class="ResultItem-bottomContainer-5 "
        >
          <p
            aria-hidden="true"
            class="MuiTypography-root MuiTypography-caption   ResultItem-bottom css-o763o0-MuiTypography-root"
          >
            Wed 8/31 general.time.short 9:00 PM
          </p>
        </div>
      </div>
    </div>
  </li>
  <li
    aria-hidden="true"
  >
    <hr
      class="MuiDivider-root MuiDivider-inset ResultItem-divider-17 css-uls1yk-MuiDivider-root"
    />
  </li>
</div>
`;
